master/internal/rm/kubernetesrm/jobs.go

```diff
@@ -0,0 +1,1869 @@
+package kubernetesrm
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	batchV1 "k8s.io/api/batch/v1"
+	"k8s.io/client-go/informers"
+	typedBatchV1 "k8s.io/client-go/kubernetes/typed/batch/v1"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/exp/maps"
+	k8sV1 "k8s.io/api/core/v1"
+	k8error "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/watch"
+	k8sClient "k8s.io/client-go/kubernetes"
+	typedV1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/homedir"
+
+	"github.com/determined-ai/determined/master/internal/config"
+	"github.com/determined-ai/determined/master/internal/db"
+	"github.com/determined-ai/determined/master/internal/rm/rmevents"
+	"github.com/determined-ai/determined/master/internal/sproto"
+	"github.com/determined-ai/determined/master/pkg/cproto"
+	"github.com/determined-ai/determined/master/pkg/device"
+	"github.com/determined-ai/determined/master/pkg/logger"
+	"github.com/determined-ai/determined/master/pkg/model"
+	"github.com/determined-ai/determined/master/pkg/set"
+	"github.com/determined-ai/determined/master/pkg/syncx/waitgroupx"
+	"github.com/determined-ai/determined/master/pkg/tasks"
+	"github.com/determined-ai/determined/proto/pkg/apiv1"
+
+	// Used to load all auth plugins.
+	_ "k8s.io/client-go/plugin/pkg/client/auth"
+)
+
+// ResourceTypeNvidia describes the GPU resource type.
+const ResourceTypeNvidia = "nvidia.com/gpu"
+
+const (
+	getAgentsCacheDuration = 15 * time.Second
+	summarizeCacheDuration = 5 * time.Second
+)
+
+type jobMetadata struct {
+	jobName      string
+	allocationID model.AllocationID
+}
+
+type jobSchedulingStateCallbackFn func(jobSchedulingStateChanged)
```

Comment:
precedent in the repo is that we name our callback functions type `XYZCallback` (no `Fn`)